qiskit/circuit/quantumcircuit.py

```diff
@@ -3323,13 +3323,15 @@ def barrier(self, *qargs: QubitSpecifier, label=None) -> InstructionSet:
         """
         from .barrier import Barrier
 
-        qubits = (
+        if qargs:
             # This uses a `dict` not a `set` to guarantee a deterministic order to the arguments.
-            list({q: None for qarg in qargs for q in self.qbit_argument_conversion(qarg)})
-            if qargs
-            else self.qubits.copy()
-        )
-        return self.append(Barrier(len(qubits), label=label), qubits, [])
+            qubits = tuple({q: None for qarg in qargs for q in self.qbit_argument_conversion(qarg)})
+            return self.append(CircuitInstruction(Barrier(len(qubits), label=label), qubits, ()))
+        else:
+            qubits = self.qubits.copy()
```

Comment:
The `copy` might not be necessary, but it does not hurt performance so I left it as it was originally coded.